nixos/modules/config/system-environment.nix

```diff
@@ -18,25 +23,63 @@ in
       default = {};
       description = ''
         A set of environment variables used in the global environment.
-        These variables will be set by PAM.
-        The value of each variable can be either a string or a list of
-        strings.  The latter is concatenated, interspersed with colon
-        characters.
+        These variables will be set by PAM early in the login process.
+        Variable substitution is available as described in
+        <citerefentry>
+          <refentrytitle>pam_env.conf</refentrytitle>
+          <manvolnum>5</manvolnum>
+        </citerefentry>.
+        In particular, the user's home directory can be accessed
+        through <code>@{HOME}</code> and the user name through
+        <code>@{PAM_USER}</code>.
+
+        Note, values may not contain the <literal>"</literal>
+        character.
+
+        The value of each session variable can be either a string or a
+        list of strings. The latter is concatenated, interspersed with
+        colon characters.
       '';
       type = with types; attrsOf (either str (listOf str));
       apply = mapAttrs (n: v: if isList v then concatStringsSep ":" v else v);
     };
 
+    environment.profileRelativeSessionVariables = mkOption {
+      type = types.attrsOf (types.listOf types.str);
+      example = { PATH = [ "/bin" ]; MANPATH = [ "/man" "/share/man" ]; };
+      description = ''
+        Attribute set of environment variable used in the global
+        environment. These variables will be set by PAM early in the
+        login process.
+
+        Each attribute maps to a list of relative paths. Each relative
+        path is appended to the each profile of
+        <option>environment.profiles</option> to form the content of
+        the corresponding environment variable.
+      '';
+    };
+
   };
 
   config = {
 
-    system.build.pamEnvironment = pkgs.writeText "pam-environment"
-       ''
-         ${concatStringsSep "\n" (
-           (mapAttrsToList (n: v: ''${n}="${concatStringsSep ":" v}"'')
-             (zipAttrsWith (const concatLists) ([ (mapAttrs (n: v: [ v ]) cfg.sessionVariables) ]))))}
-       '';
+    environment.sessionVariables =
```

Comment:
With the existing `environment.variables` and `environment.profileRelativeEnvVars` the former is always first. Indeed the code never injects the entries generated from `environment.profileRelativeEnvVars` into `environment.variables`.

I actually did the same here before but changed to this method since I figured it was a bit more elegant. Might be safer to revert this change and use the same setup as for `environment.profileRelativeEnvVars`.

Reply:
I've reworked the code a bit and now the semantics is the same as for `variables` and `profileRelativeEnvVars`. That is, the values in `sessionVariables` will end up in front of the ones introduced in `profileRelativeSessionVariables`.